bpf: switch csum_l4_replace and ipv4_dec_ttl to csum_diff

borkmann opened the pull request:
Moved out of #10491, since it is blocked by #10517.

Another option is to drop it and implement the 2 byte diff in XDP at the expense of more complexity. Need to double check on 4.19.57+ kernels.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/10521)
<!-- Reviewable:end -->

borkmann committed:
bpf: skip policy call map if not defined

Extend ifdef SKIP_POLICY_MAP wrapper around inline functions that
use patterns like tail_call(ctx, &POLICY_CALL_MAP, ep->lxc_id) in
order to avoid compile issues for files setting SKIP_POLICY_MAP to 1.

This is needed for XDP layer. sock LB and sockmap programs already
set SKIP_POLICY_MAP today, but they didn't run into this issue since
they didn't include l3.h.

We already define the map today as:

  #ifndef SKIP_POLICY_MAP
  /* Global map to jump into policy enforcement of receiving endpoint */
  struct bpf_elf_map __section_maps POLICY_CALL_MAP = {
        .type           = BPF_MAP_TYPE_PROG_ARRAY,
        .id             = CILIUM_MAP_POLICY,
        .size_key       = sizeof(__u32),
        .size_value     = sizeof(__u32),
        .pinning        = PIN_GLOBAL_NS,
        .max_elem       = POLICY_PROG_MAP_SIZE,
  };
  #endif /* SKIP_POLICY_MAP */

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

qmonnet commented on bpf/lib/csum.h:
```diff
@@ -62,7 +62,13 @@ static __always_inline int csum_l4_replace(struct __ctx_buff *ctx, int l4_off,
 					   __be32 from, __be32 to, int flags)
 {
-	return l4_csum_replace(ctx, l4_off + csum->offset, from, to, flags | csum->flags);
+	__be32 sum = csum_diff(&from, sizeof(from), &to, sizeof(to), 0);
```
Nit: You could maybe call `csum_diff_small()` here and avoid to type the two `sizeof()`. Not that it matters much.

qmonnet reviewed (approved):
Looks good!

borkmann committed:
bpf: implement __csum_replace_by_4 for xdp context

Rather than converting few places that update ports to use extra
csum_diff helper call, just implement support for it inline in XDP.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: fix a few compile errors from xdp with __maybe_unused

In XDP context these function parameters are not used hence mark
them as such.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: add HAVE_PROG_TYPE_HELPER feature mapper for use in skb/xdp

Out of our main code, we cannot randomly test via HAVE_PROG_TYPE_HELPER()
since some features are only in skb context and some are only in xdp
context and while they are available in skb context, we cannot compile
them in for xdp. Hence generalize the feature mapper and include only
what is relevant for a given context.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
